components/suggest/src/keyword.rs

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/// Given a list of keywords for a suggestion, returns a phrase that best
+/// completes the user's query. This function uses two heuristics to pick the
+/// best match:
+///
+/// 1. Find the first keyword in the list that has at least one more word than
+///    the query, then trim the keyword up to the end of that word.
+/// 2. If there isn't a keyword with more words, pick the keyword that forms the
+///    longest suffix of the query. This might be the query itself.
+pub fn full_keyword(query: &str, keywords: &[impl AsRef<str>]) -> String {
+    let query_words_len = query.split_whitespace().count();
+    let min_phrase_words_len = if query.ends_with(char::is_whitespace) {
+        // If the query ends with a space, find a keyword with at least one more
+        // word, so that the completed phrase can show a word after the space.
+        query_words_len + 1
+    } else {
+        query_words_len
+    };
+    keywords
+        .iter()
+        .map(AsRef::as_ref)
+        .filter(|phrase| phrase.starts_with(query))
+        .map(|phrase| phrase.split_whitespace().collect::<Vec<_>>())
```

Comment:
what do you think about using `filter_map` with `then_some` here and merge those three? I'm conflicted whether that would make it easier or harder to read 🤔 Something like: (although my indentation is off)
```suggestion
        .filter_map(|phrase| phrase
                              .as_ref()
                              .starts_with(query)
                              .then_some(
                                 phrase
                                   .as_ref()
                                   .split_white_space()
                                   .collect::<Vec<_>>()
                               )
                 )
```

Reply:
Oh, thanks for reminding me about `then` and `then_some`! I wrote it with `filter_map` at first, but with the `if` check written out. I think the extra level of indentation and nested closure still make it a little harder to read, though:

```rust
.filter_map(|phrase| {
    phrase
        .as_ref()
        .starts_with(query)
        .then(|| phrase.as_ref().split_whitespace().collect::<Vec<_>>())
})
```

(I used `then` instead of `then_some` to avoid eagerly evaluating `split_whitespace`).

Would you be OK with keeping it as-is? There's a good chance we'll end up dropping this function in favor of using `full_keywords` from the Remote Settings data, so this might end up going away.

I did love getting to use Rust's iterators for these, and was impressed with how well they were able to follow the [English description](https://searchfox.org/mozilla-central/rev/c615dc4db129ece5cce6c96eb8cab8c5a3e26ac3/browser/components/urlbar/private/AdmWikipedia.sys.mjs#209-231)!